### ISSUE TITLE
Error 70001

### ISSUE THREAD

**Issue by bil-ash** [Softmotions/iowow#25]
I am using ejdb2_node. At times(not always), I get error while closing the database. The error boils down to an error in iowow with error code 70001. Could you provide details about the error and if any workaround exists.

![Annotation 2019-09-27 121236](https://user-images.githubusercontent.com/33216854/65747872-3079c300-e120-11e9-9410-d56549c05382.png)

**Comment by adamansky** [comment]
Hi! Thank you for report. Could you please show me part of js code which closes the database?

**Comment by adamansky** [comment]
Also, if it is possible, could please distille a minimal test application which may reproduce this fail then send me to adamansky@gmail.com?  It it will significantly speed up investigation for this case.

**Comment by bil-ash** [comment]
The problem persists even after upgrading to latest version(1.0.14). Attaching herewith the application which is giving the error.
[test.zip](https://github.com/Softmotions/iowow/files/3662857/test.zip)

**Comment by bil-ash** [comment]
The problem persists even after upgrading ejdb2_node to latest version(1.0.14). I have come across an interesting situation. If I delete the database file and the wal file and then run the program, there is no error on the first run, errors start only from the second time the program is run. I would also like to mention that I maintained a gap of 30 seconds between two subsequent runs to ensure that there is sufficient time between database closing and opening. Attaching herewith my program.
[test.zip](https://github.com/Softmotions/iowow/files/3663008/test.zip)

**Comment by adamansky** [comment]
Cannot reproduce on:
```
OS: Ubuntu 18.04.3 LTS
node --version
v10.16.3
```
Please send me your nodejs version + os version used

**Comment by adamansky** [comment]
It seems to be mmapping issue tightly related to your environment, so it will be great to know as much as possible about os (version, cpu, kernel), c compiler version, nodejs version.

**Comment by bil-ash** [comment]
Ubuntu 16.04.4 LTS x64
node 10.16.3
The C compiler identification is GNU 5.4.0

**Comment by adamansky** [comment]
This error means that some system call, I suppose `mmap` or `pwrite` sets `errno` to `EINVAL` on database close.

I cannot reproduces this issue. Take a look what I did:

```sh
docker pull ubuntu:16.04
docker run --name ejdb_test_25 -it ubuntu:16.04

apt-get update && apt-get install -y cmake build-essential git curl unzip wget

curl -sL https://deb.nodesource.com/setup_10.x | bash -
curl -sL https://dl.yarnpkg.com/debian/pubkey.gpg | apt-key add -
echo "deb https://dl.yarnpkg.com/debian/ stable main" | tee /etc/apt/sources.list.d/yarn.list

apt-get update &&  apt-get install -y yarn nodejs

mkdir ejdb2_test_25 && cd ./ejdb2_test_25

wget  https://github.com/Softmotions/iowow/files/3663008/test.zip
unzip test.zip

yarn init 
yarn add ejdb2_node


# Then run many times, no error
node ./ej.js 
```

**Comment by bil-ash** [comment]
Could you please tell which version of C compiler is installed on your docker system ,mine is 5.4.0 as mentioned above. Who knows even difference in C compiler version might lead to errors.

**Comment by adamansky** [comment]
Exactly 
```
gcc (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
```
Okay, give me few days, I will inject more verbose logging around problem code, then will ask you to run it again.

**Comment by bil-ash** [comment]
okay

**Comment by adamansky** [comment]
Hi! Sorry for a long delay I'm was very busy on work.
Could you please remove `ejdb2_node` package then install `@softmotions/ejdb2_ticket_25` and sen me console output of test program. Thank you!

**Comment by bil-ash** [comment]
I am attaching herewith the pics of running the program in 2 sets of 3 times in a row. The results were identical in both the cases. The first run was OK(only the later part of the console messages is captured in the pic). The second run led to a **put** error. So I had to use Ctrl+C to force end the program. The third run led to **close** error, most probably due to put error/force close in 2nd run.

I would also like to inform you that I have switched to sqlite for my requirement. I had to make a few APIs and [strapi](https://strapi.io) provided a set of utilities in addition to the ability to use embedded database(sqlite) which is perfect for me. So I may not _be able_/ _wish_ to provide feedback in the future on this issue.
![snap1](https://user-images.githubusercontent.com/33216854/66838430-47bf0a00-ef82-11e9-826a-e7503e79d69e.png)
![snap2](https://user-images.githubusercontent.com/33216854/66838431-47bf0a00-ef82-11e9-97e9-e5f31cb08dbf.png)
